Add new functions to Opal

theiostream opened the pull request:
Add some new functions to Opal which are required by WebKit.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/gnustep/libs-opal/7)
<!-- Reviewable:end -->

theiostream committed:
CGContext: correctly export CGContextDrawShading() stub

In CGContext.m we had CGContextDrawShading()'s definition repeated from
the header; we did not, however, have an actual stub for this function,
which caused linking errors with Opal clients. This turns this still
unimplemented function into an actual stub.

theiostream committed:
CGContext: properly implement ellipse-related functions

Add an actual implementation for CGContextAddEllipseInRect() which makes
use of CGPath APIs. Implement CGContextFillEllipseInRect() and
CGContextStrokeEllipseInRect() as well.

theiostream committed:
CGContext: define some private getter functions

CoreGraphics posesses the private getter counterparts for some setter
functions, such as antialiasing and font smoothing. This implements
those getters.

theiostream committed:
CGContext: define CGContextGetType()

Define the private CoreGraphics CGContextGetType() API, which determines
the type of a CGContextRef (a bitmap context, PostScript context, PDF
context etc.).

theiostream committed:
CGContext: implement CGContextSetCTM()

Implement the private function CGContextSetCTM(), which allows clients
to set the current transformation matrix of a context.

ivucica commented on Headers/CoreGraphics/CGContext.h:
```diff
@@ -562,6 +578,19 @@ CGRect CGContextConvertRectToDeviceSpace(CGContextRef ctx, CGRect rect);
+bool CGContextPathContainsPoint(CGContextRef c,
+    CGPoint point, CGPathDrawingMode mode);
+
+/* CoreGraphics Private APIs */
```
Since these are not part of the public API, I propose we move them into a separate header + implementation file.

fredkiefer replied:
This function gets exported by the Apple implementation as well. Having it here is fine.

ivucica replied:
Well, I selected line 584 when I was creating the commit, not my fault that GitHub has... flaws. :smile:

I was referring to all the APIs below that were marked as private. I did not check any to be private except `CGContextSetCTM()`

ivucica reviewed (commented):
I can merge as is, but I'd really like private stuff slightly more ... distinct.

fredkiefer commented on Headers/CoreGraphics/CGContext.h:
```diff
@@ -562,6 +578,19 @@ CGRect CGContextConvertRectToDeviceSpace(CGContextRef ctx, CGRect rect);
+
+CGContextType CGContextGetType(CGContextRef ctx);
+
+void CGContextSetCTM(CGContextRef ctx, CGAffineTransform m);
```
The last two are the one functions I cannot find in the Apple header, these should be moved away.

ivucica replied:
Out of these, it looks like only `CGContextGetShouldAntialias()` is public.

fredkiefer commented on Headers/CoreGraphics/CGContext.h:
```diff
@@ -138,6 +138,22 @@ enum {
 };
 typedef int CGTextEncoding;
 
+// NOTE: This `typedef enum` in opposition to defining enum constants and
```
Where did you find these values?  As WebKit has its own definitions of these values we should kept them private within Opal.

fredkiefer reviewed (commented):
I agree with Ivan that we should keep the private functions and types within Opal and not export them. Please changes this and the code could be merged from my side.

theiostream committed:
CGContext: create CGContext private function stubs

Create stubs for yet unimplemented private CoreGraphics functions, for
increased compatibility with CG.

theiostream committed:
CGFont: implement stub for CGFontGetGlyphAdvancesForStyle

Create a stub for the private function CGFontGetGlyphAdvancesForStyle
for increased compatibility.

theiostream committed:
CGImageSource: add private stubs

Add private constant and function stubs for CGImageSource for increased
compatibility.

theiostream committed:
CTFont: add private function stubs

Add stubs for some private functions and constants in CTFont for
compatibility purposes.

theiostream committed:
CTFontDescriptor: add stub for CTFontDescriptorIsSystemUIFont()

Create a stub for the CTFontDescriptorIsSystemUIFont() private function
for increased compatibility.

theiostream committed:
CTLine: add stub for CTLineCreateWithUniCharProvider()

Add a stub for the CTLineCreateWithUniCharProvider() private function
for increased compatibility purposes.

theiostream committed:
CTRun: add private function stubs

Add stubs for private functions in CTRun for increased compatibility.

theiostream committed:
CTTypesetter: add stub for CTTypesetterCreateWithUniCharProviderAndOptions()

Add stub for CTTypesetterCreateWithUniCharProviderAndOptions(), a
private function, for increased compatibility.

fredkiefer commented:
Please move all the private functions out of the public headers. Otherwise we won't be able to merge this.

ivucica commented:
Will there be more progress on this? :) @theiostream

fredkiefer commented:
@ivucica I suggest we just merge this pull request. I am not totally happy with it, but there won't be any future progress and at least there isn't anything fundamentally wrong in this code.

ivucica commented:
Marking this as something to review one of the evenings this week. I see the old comments about not making some internal API public, and we should at least do that when merging, otherwise we'll never do it.
